### ISSUE TITLE
Why is KL decreasing faster with default parameters compared to the FIt-SNE implementation?

### ISSUE THREAD

**Issue by dkobak** [pavlin-policar/openTSNE#80]
I was under the impression that all important optimization parameters used in openTSNE are the same as in FIt-SNE:

```
class openTSNE.TSNE(n_components=2, perplexity=30, learning_rate=200, 
early_exaggeration_iter=250, early_exaggeration=12, n_iter=750, exaggeration=None, theta=0.5, 
n_interpolation_points=3, min_num_intervals=50, ints_in_interval=1, initialization='pca', 
metric='euclidean', metric_params=None, initial_momentum=0.5, final_momentum=0.8, 
min_grad_norm=1e-08, max_grad_norm=None, n_jobs=1, neighbors='approx', 
negative_gradient_method='fft', callbacks=None, callbacks_every_iters=50, random_state=None)
```

and

```
def fast_tsne(X, theta=.5, perplexity=30, map_dims=2, max_iter=1000, 
              stop_early_exag_iter=250, K=-1, sigma=-1, nbody_algo='FFT', knn_algo='annoy',
              mom_switch_iter=250, momentum=.5, final_momentum=.8, learning_rate=200,
              early_exag_coeff=12, no_momentum_during_exag=False, n_trees=50, 
              search_k=None, start_late_exag_iter=-1, late_exag_coeff=-1,
              nterms=3, intervals_per_integer=1, min_num_intervals=50,            
              seed=-1, initialization=None, load_affinities=None,
              perplexity_list=None, df=1, return_loss=False, nthreads=None)
```

However, I have just noticed when using one particular dataset that KL decreases faster when I use openTSNE. When I run this:

```
Z = fast_tsne(X, perplexity=30, seed=42)
```

I get 

```
Iteration 50 (50 iterations in 0.91 seconds), cost 6.444297
Iteration 100 (50 iterations in 0.87 seconds), cost 6.270150
Iteration 150 (50 iterations in 0.90 seconds), cost 4.691937
Iteration 200 (50 iterations in 0.88 seconds), cost 4.266439
Iteration 250 (50 iterations in 0.86 seconds), cost 4.019699
Unexaggerating Ps by 12.000000
Iteration 300 (50 iterations in 0.89 seconds), cost 3.404685
Iteration 350 (50 iterations in 0.91 seconds), cost 3.043569
Iteration 400 (50 iterations in 0.87 seconds), cost 2.794367
Iteration 450 (50 iterations in 0.92 seconds), cost 2.613639
Iteration 500 (50 iterations in 0.93 seconds), cost 2.471313
Iteration 550 (50 iterations in 1.14 seconds), cost 2.356686
Iteration 600 (50 iterations in 1.28 seconds), cost 2.271862
Iteration 650 (50 iterations in 1.47 seconds), cost 2.190770
Iteration 700 (50 iterations in 1.68 seconds), cost 2.127095
Iteration 750 (50 iterations in 1.98 seconds), cost 2.074458
Iteration 800 (50 iterations in 2.50 seconds), cost 2.027436
Iteration 850 (50 iterations in 2.42 seconds), cost 1.981508
Iteration 900 (50 iterations in 2.63 seconds), cost 1.940266
Iteration 950 (50 iterations in 3.39 seconds), cost 1.915046
Iteration 1000 (50 iterations in 3.72 seconds), cost 1.875520
Wrote the 23822 x 2 data matrix successfully.
```

whereas when I run

```
Zo = TSNE(initialization='random', random_state=42, callbacks=ErrorLogger(), n_jobs=-1).fit(X)
```

I get

```
Iteration   50, KL divergence  6.0437, 50 iterations in 1.6216 sec
Iteration  100, KL divergence  4.5617, 50 iterations in 1.6222 sec
Iteration  150, KL divergence  4.2162, 50 iterations in 1.6578 sec
Iteration  200, KL divergence  4.0487, 50 iterations in 1.7035 sec
Iteration  250, KL divergence  3.9388, 50 iterations in 1.6815 sec
Iteration   50, KL divergence  3.1818, 50 iterations in 1.6840 sec
Iteration  100, KL divergence  2.7344, 50 iterations in 1.8720 sec
Iteration  150, KL divergence  2.4542, 50 iterations in 1.9890 sec
Iteration  200, KL divergence  2.2686, 50 iterations in 2.1730 sec
Iteration  250, KL divergence  2.1360, 50 iterations in 2.3945 sec
Iteration  300, KL divergence  2.0377, 50 iterations in 2.6713 sec
Iteration  350, KL divergence  1.9628, 50 iterations in 2.9563 sec
Iteration  400, KL divergence  1.9047, 50 iterations in 3.4022 sec
Iteration  450, KL divergence  1.8599, 50 iterations in 3.6570 sec
Iteration  500, KL divergence  1.8257, 50 iterations in 3.9738 sec
Iteration  550, KL divergence  1.7997, 50 iterations in 4.6501 sec
Iteration  600, KL divergence  1.7799, 50 iterations in 5.0196 sec
Iteration  650, KL divergence  1.7647, 50 iterations in 5.3889 sec
Iteration  700, KL divergence  1.7527, 50 iterations in 5.7517 sec
Iteration  750, KL divergence  1.7420, 50 iterations in 5.6339 sec
```

and indeed with FIt-SNE I get the embedding that spans roughly from -50 to 50, whereas with openTSNE it spans roughly from -80 to 80. I would expect this to happen if openTNSE used higher learning rate, but it's set to 200 in both implementations.

**Comment by pavlin-policar** [comment]
It's possible that either implementation skipped some constants. When doing minimization the constants don't change the objective landscape other than scaling things up and down, so I (or George) might have left out some constants to save on the math operations during optimizations.

My jumping off point was scikit-learn, and I really only looked at FIt-SNE when I implemented the fast approximation for the gradients. But you're right, this is definitely confusing and it might be worth going through the exact math and making sure all the constants are in place in both implementations.

**Comment by dkobak** [comment]
Oh, I see. This would make sense. FIt-SNE used the same conventions as Maaten's original BH t-SNE implementation, but I never checked if scikit-learn agrees with them. It can very well be that one needs to scale the learning rate up/down by some constant conversion factor when comparing them... We should check.

**Comment by dkobak** [comment]
There is a factor 4 in the t-SNE gradient and Maaten's implementation omitted it. If scikit-learn and openTSNE used the "correct" gradient with factor 4, then the learning rates will differ by a factor of 4.

This would actually be rather unfortunate because it makes "learning rate" an ambiguous value.

**Comment by pavlin-policar** [comment]
That's quite unfortunate and, as you said, makes the learning rate arbitrary. I am not entirely sure what the best course of action here is because, on the one hand, our gradients are "correct" in terms of the math and on the other hand, it's inconsistent with Maaten's implementation, FIt-SNE and probably MulticoreTSNE.

It might be most practical to drop the "4" simply because if anybody takes a closer look at t-SNE and actually wants to experiment with it, it seems more likely they start from Maaten's implementation and not scikit-learn. I'd love to get your thoughts on this.

**Comment by dkobak** [comment]
Before discussing what to do with it, let's make sure that this is the actual reason. That there is no "4" term in the Maaten's implementation, I know for sure (see footnote 13 in https://arxiv.org/abs/1902.05804). Are we certain that scikit and openTSNE, both have it?

**Comment by pavlin-policar** [comment]
openTSNE has it [here](https://github.com/pavlin-policar/openTSNE/blob/master/openTSNE/tsne.py#L1233) and scikit-learn [here](https://github.com/scikit-learn/scikit-learn/blob/7813f7efb/sklearn/manifold/t_sne.py#L259). The equation comes from _Learning a Parametric Embedding by Preserving Local Structure_ by VDM.

**Comment by dkobak** [comment]
OK you are right, this must be it.

>  I am not entirely sure what the best course of action here is because, on the one hand, our gradients are "correct" in terms of the math and on the other hand, it's inconsistent with Maaten's implementation, FIt-SNE and probably MulticoreTSNE.

Yeah, it's screwed up... I am not sure either. George won't want to change it in FIt-SNE because he wants to maintain consistency with Maaten's implementations (his BH implementation became quite standard, it's used in R, etc.; I'm actually not sure what Matlab uses here https://www.mathworks.com/help/stats/tsne.html -- it might be worth a check). On the other hand, I am sure scikit-learn people will not want to drop the 4 either, because it wouldn't be self-consistent anymore.

So unfortunately some inconsistency is here to stay, and the question for openTSNE is only if you want to adopt Maaten's convention or scikit-learn convention.

> It might be most practical to drop the "4" simply because if anybody takes a closer look at t-SNE and actually wants to experiment with it, it seems more likely they start from Maaten's implementation and not scikit-learn.

I agree. I think this would be my preference too. Even though I don't like that this 4 factor is just arbitrarily "lost"...

In any case, it's important for me because in our preprint we follow Belkina et al. in recommending to set learning rate = n/12, but this refers to the Maaten's learning rate (at least I *think* that that's what Belkina et al. use!). It translates to learning rate = n/48 in scikit and currently in openTSNE... Of course this 1/12 factor is largely arbitrary; it can be that n/3 would work as well.

> openTSNE has it here and scikit-learn here.

Waaait, both have `dof` parameter?! Does it mean one could actually change it? I never saw it in scikit-learn (or openTSNE) docs.

**Comment by linqiaozhi** [comment]
> Yeah, it's screwed up... I am not sure either. George won't want to change it in FIt-SNE because he wants to maintain consistency with Maaten's implementations (his BH implementation became quite standard, it's used in R, etc.;

I agree with you both that it is unfortunate. Dmitry accurately characterized my position; I hesitate to make the implementation inconsistent with Van der Maaten's. Also, since FIt-SNE currently does not include the 4, I think it would become even more confusing if it suddenly changes to have the 4. 

However, I would definitely be open to adding a note about this to the documentation. Perhaps the description of the learning rate parameter would be a good place, this way anyone who is looking to change the learning rate will see a note that the implementation does or does not include the constant 4. 

> In any case, it's important for me because in our preprint we follow Belkina et al. in recommending to set learning rate = n/12, but this refers to the Maaten's learning rate (at least I think that that's what Belkina et al. use!). It translates to learning rate = n/48 in scikit and currently in openTSNE... Of course this 1/12 factor is largely arbitrary; it can be that n/3 would work as well.

I agree. I think the reason that recommendation is helpful is because it makes the learning rate scale with n; I would be surprised if the constant matters a great deal.

**Comment by pavlin-policar** [comment]
> Waaait, both have dof parameter?! Does it mean one could actually change it? I never saw it in scikit-learn (or openTSNE) docs.

So scikit-learn has it in code, but it's not a tunable parameter in the public API. But the implementation is there for Barnes-Hut. Also in openTSNE. I never got around to adding it to George's approximation, so I, too, left it out of the public API. Again, I've been meaning to add this, but I've been really busy finishing up my master's thesis, and a whole lot of other things have piled up :/

One more note I'd make here is that in Maaten's follow-up paper on parametric t-SNE, which introduces the variable degrees of freedom, this also shows up in the gradient:

![image](https://user-images.githubusercontent.com/5758119/61991618-fb36f300-b052-11e9-8fb1-7ae14e8410e7.png)

This is what's used in both scikit-learn and openTSNE, and if we allow changing the degrees of freedom, this would impact the gradient. So then the learning rate wouldn't absorb the constant "4" factor, but a variable factor, making things even more confusing. I'm worried that this variable factor might affect which learning rates would work.

**Comment by dkobak** [comment]
> This is what's used in both scikit-learn and openTSNE, and if we allow changing the degrees of freedom, this would impact the gradient. So then the learning rate wouldn't absorb the constant "4" factor, but a variable factor, making things even more confusing. I'm worried that this variable factor might affect which learning rates would work.

Yes. We did not have this problem when implementing varying degree of freedom in FIt-SNE because we used a slightly different kernel parametrization and there the factor is `4` independent of the degree of freedom -- see Appendix in https://arxiv.org/pdf/1902.05804.pdf.

**Comment by dkobak** [comment]
Hey Pavlin, did you decide on whether you are keeping the factor 4 or not? I am preparing the Art of t-SNE paper for the final submission and would like to mention this issue in the methods. I'll say that FIt-SNE followed BH t-SNE in not using factor 4, but scikit-learn does use it. I could mention openTSNE but I need to know where to put it :-)

**Comment by pavlin-policar** [comment]
Hey, I'll be removing it to stay consistent with Maaten's and George's implementations. It makes more sense to be consistent with that than with scikit-learn. Unfortunately, I haven't gotten around to it yet.

**Comment by dkobak** [comment]
OK, great. I will mention it in as such then. As long as you implement it during the next several weeks (before the paper in its final form is published), nobody will get confused :)

**Comment by pavlin-policar** [comment]
@dkobak The gradients should now be more or less consistent with FIt-SNE.
